src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3152,6 +3153,312 @@ void MacroAssembler::check_klass_subtype(Register sub_klass,
+  // current slot (at secondary_supers[r_array_index]) has not yet
+  // been inspected, and r_array_index may be out of bounds if we
+  // wrapped around the end of the array.
+
+  { // This is conventional linear probing, but instead of terminating,
+    // when a null entry is found in the table, we maintain a bitmap
+    // in which a 0 indicates missing entries.
+    // The check above guarantees there are 0s in the bitmap, so the loop
+    // eventually terminates.
+
+#ifdef ASSERT
+    // We should only reach here after having found a bit in the bitmap.
+    // Invariant: array_length == popcount(bitmap)
+    z_ltgr(r_array_length, r_array_length);
+    asm_assert(bcondHigh, "array_length > 0, should hold", 22);
+#endif
+
+    // Compute limit in r_array_length
+    add2reg(r_array_length, -1);
+    z_sllg(r_array_length, r_array_length, LogBytesPerWord);
+
+    NearLabel L_loop;
+    bind(L_loop);
+
+    // Check for wraparound.
+    z_cgr(r_array_index, r_array_length);
+    z_locgr(r_array_index, r_result, bcondHigh); // r_result is containing 0, so we're good :)
+
+    z_cg(r_super_klass, Address(r_array_base, r_array_index));
+    z_bre(L_done); // success
+
+    // look-ahead check (Bit 2), if bit-2 is also 0, we're done
+    testbit(r_bitmap, 2);
+    z_bfalse(L_failure);
+
+    z_rllg(r_bitmap, r_bitmap, 64-1); // rotate right
+    add2reg(r_array_index, BytesPerWord);
+
+    z_bru(L_loop);
+  }
+
+  { // Degenerate case: more than 64 secondary supers.
+    // FIXME: We could do something smarter here, maybe a vectorized
+    // comparison or a binary search, but is that worth any added
+    // complexity?
+
+    bind(L_huge);
+    repne_scan(r_array_base, r_super_klass, r_array_length, r_result);
+
+    z_bru(L_done); // whatever we got from repne_scan, forward it
+  }
+
+  bind(L_failure);
+  z_lghi(r_result, 1);
+
+  bind(L_done);
+  BLOCK_COMMENT("} lookup_secondary_supers_table_slow_path");
+}
+
+// Make sure that the hashed lookup and a linear scan agree.
+void MacroAssembler::verify_secondary_supers_table(Register r_sub_klass,
+                                                   Register r_super_klass,
+                                                   Register r_result /* expected */,
+                                                   Register r_temp1,
+                                                   Register r_temp2,
+                                                   Register r_temp3) {
+  assert_different_registers(r_sub_klass, r_super_klass, r_result, r_temp1, r_temp2, r_temp3);
+
+  const Register
+    r_array_base   = r_temp1,
+    r_array_length = r_temp2,
+    r_array_index  = r_temp3,
+    r_bitmap       = noreg; // unused
+
+  const Register r_one = Z_R0_scratch;
+  z_lghi(r_one, 1); // for locgr down there, to a load result for failure
+
+  LOOKUP_SECONDARY_SUPERS_TABLE_REGISTERS;
+
+  BLOCK_COMMENT("verify_secondary_supers_table {");
+
+#ifdef ASSERT
+  {
+    // r_result should have either 0 or 1 value
```

Comment:
What is this assert block for? `result` is either zero or nonzero, and there's no need for anything stronger.

Reply:
Removed.